### PR TITLE
Step5 - Car Racing (Refactoring)

### PR DESCRIPTION
**Pull request by krrong**
Hello Suhyeon, It's already final step of car racing mission, which is a bit sad.
The comments that you left in previous steps were really valuable and helpful for me!
I'm looking forward to the feedback on this step as well. Thank you for your help! 😆

### CODE REVIEW

File: src/main/kotlin/racingcar/domain/car/Winners.kt [hunk 1]
```diff
@@ -1,3 +1,10 @@
 package racingcar.domain.car
 
-data class Winners(val cars: List<Car>)
+data class Winners(val cars: List<Car>) {
+    companion object {
+        fun of(cars: List<Car>): Winners {
+            val maxPosition = cars.maxOf { it.position }
+            return Winners(cars.filter { it.position == maxPosition })
+        }
+    }
```
Comment (krrong) [inline src/main/kotlin/racingcar/domain/car/Winners.kt:9]: Delegated the logic to get winners from Cars class to Winners class and placed it in companion object.

**Review by krrong — COMMENTED** [review]
https://github.com/next-step/kotlin-racingcar/pull/1808#discussion_r2020865865 

I supposed to leave a comment that makes you understand my intention but I forgot it.
I tried to create secondary constructor but `List<Name>` and `List<Car>` are the same signatures for JVM so it was not available. 🤔

File: README.md
Comment (wisemuji) [inline README.md:1]: > https://github.com/next-step/kotlin-racingcar/pull/1808#discussion_r2020865865
I supposed to leave a comment that makes you understand my intention but I forgot it.
I tried to create secondary constructor but List<Name> and List<Car> are the same signatures for JVM so it was not available. 🤔

I understand! btw, the article I referenced shows workarounds for JVM signature duplicates 🙂

File: README.md [hunk 2]
```diff
@@ -13,6 +13,10 @@
 - Car names should be comma-separated when entered by the user.
 - After the race is complete, display the winners. There can be multiple winners if they reach the same final position.
 
+## Step 5 Requirements
+- Implement objects containing core business logic in the domain package and UI-related objects in the view package.
+- Refactor using the MVC pattern, ensuring that objects in the view package can depend on objects in the domain package, but not the other way around.
```
Comment (wisemuji) [inline README.md:18]: Since you already met Step 5 requirements in the previous step, there were no critical feedbacks on the step!

File: src/main/kotlin/racingcar/domain/car/Cars.kt [hunk 3]
```diff
@@ -9,8 +9,7 @@ class Cars(val cars: List<Car>) {
     }
 
     fun getWinners(): Winners {
-        val maxPosition = cars.maxOf { it.position }
-        return Winners(cars.filter { it.position == maxPosition })
+        return Winners.of(cars)
```
Comment (wisemuji) [inline src/main/kotlin/racingcar/domain/car/Cars.kt]: Since this could be treated as a property like below, which do you prefer? 🤔 
```kotlin
val winners: Winners get() = Winners.of(cars)
```

Maybe the video below can help you organize your thoughts!
https://www.youtube.com/watch?v=R38bW3AItf8

File: src/main/kotlin/racingcar/domain/car/Winners.kt (hunk 1 above)
Comment (wisemuji) [reply src/main/kotlin/racingcar/domain/car/Winners.kt:9]: Nice work 👍 

The book [[KR] The Essence of Object-Orientation](https://product.kyobobook.co.kr/detail/S000001628109) describes "message" as the following:

> One object accesses another by sending a message (message-transmission mechanism: the only way an object can access another object)
The fact that an object may receive a message means that the object may perform a responsibility corresponding to the message.
The outside object is naturally separated from the outside and the inside of the object because it can only be seen with respect to the message and not the inside of the object.
A message consists of a combination of recipient, message name, and factor.

You respected the object's responsibility and set a way to send a proper message. 👍

File: src/test/kotlin/racingcar/domain/car/WinnersTest.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package racingcar.domain.car
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import racingcar.domain.Name
+
+class WinnersTest {
+    @Test
+    fun `car1 and car2 are winners when car1 has position 1 and car2 has position 1`() {
```
Comment (wisemuji) [inline src/test/kotlin/racingcar/domain/car/WinnersTest.kt]: How about applying `ParameterizedTest` in this scenario, and testing boundary conditions?

https://www.linkedin.com/advice/3/how-can-you-test-edge-cases-boundary-conditions-nwaoc

**Review by wisemuji — CHANGES_REQUESTED** [review]
Hi Seokjin!

You did a good job with the implementation.
Please review the comment I left for you.

I have no critical feedback, so I think this PR can be merged in the next review request. ☺️

File: README.md
Comment (krrong) [reply README.md:1]: From my understanding, to use factory method is one of the workarounds.
Please correct me If I'm wrong 😊

File: src/main/kotlin/racingcar/domain/car/Cars.kt (hunk 3 above)
Comment (krrong) [reply src/main/kotlin/racingcar/domain/car/Cars.kt]: It was really valuable comment and helpful video!

File: src/test/kotlin/racingcar/domain/car/WinnersTest.kt (hunk 4 above)
Comment (krrong) [reply src/test/kotlin/racingcar/domain/car/WinnersTest.kt]: Applied `ParameterizedTest` but still curious is there any boundary conditions 🤔

File: src/test/kotlin/racingcar/domain/car/WinnersTest.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        val actual = Winners.of(cars)
+
+        // then
+        assertThat(actual.cars).hasSameElementsAs(expectedWinners)
+    }
+
+    companion object {
+        @JvmStatic
+        fun provideCarsAndWinners(): Stream<Arguments> {
+            return Stream.of(
```
Comment (wisemuji) [inline src/test/kotlin/racingcar/domain/car/WinnersTest.kt:27]: Kotlin developers might find `List` to be more familiar 🙂 
```suggestion
        fun provideCarsAndWinners(): List<Arguments> {
            return listOf(
```

File: src/test/kotlin/racingcar/domain/car/WinnersTest.kt (hunk 4 above)
Comment (wisemuji) [reply src/test/kotlin/racingcar/domain/car/WinnersTest.kt]: Nice work with `ParameterizedTest`, and I was thinking of such edge cases:
1. Empty list
2. All cars are located at zero
3. Cars with negative locations
4. Multiple cars with the same name
5. Only one car

**Review by wisemuji — APPROVED** [review]
Hi Seokjin! You really nailed it. Great job!

It was really nice to share thoughts and learn some new perspectives.
I hope it helped you improve as a developer, and now we can end this mission.

I'll be cheering you on as you work on finishing your other missions as well. 💪
